File: 反转链表/reverse_list.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "list.h"

void reverse_list(list *li) {
	node *cur = li->head;
	node *pre = NULL;
	node *next = cur->next;
	
	while (cur) {
		if (pre != NULL) {
			cur->next = pre;
		}
		pre = cur;
		cur = next;
		if (next) {
			next = next->next;
		}
	}
	li->head = pre;
}

int main(int argc, char *argv[]) {
	list *li = (list *)malloc(sizeof(list));
	node *head = (node *)malloc(sizeof(node));
	head->val = 1;
	node *second = (node *)malloc(sizeof(node));
	second->val = 2;
	node *third = (node *)malloc(sizeof(node));
	third->val = 3;
	
	head->next = second;
	//second->next = third;
	li->head = head;
	
	reverse_list(li);
	
	printf("%d\n", li->head->val);
}

File: 反转链表/ReverseList.java
class List {
	public Node head;
	
	private Node internalReverse(Node node) {
		Node cur = node;
		if (node == null) {
			return cur;
		}
		if (cur.next == null) {
			head = cur;
		} else {
			Node tail = internalReverse(cur.next);	
			tail.next = cur;
		}
		cur.next = null;
		return cur;
	}
	
	public void reverse() {
		internalReverse(head);	
	}

	public void print() {
        Node cur = head;
        while (cur != null) {
            System.out.println(cur.val);
            cur = cur.next;
        }
    }	
}

class Node {
	public int val;
	public Node next;
}

public class ReverseList {
	public static void main(String[] args) {
		Node head = new Node();
        Node first = new Node();
        Node second = new Node();
        head.val = 1;
        first.val = 2;
        second.val = 3;
        head.next = first;
        first.next = second;
        List list = new List();
        list.head = head;
        list.reverse();
        list.print();		
	}
}
